File: src/21-algoritmos-aproximacao/04-bin-packing/bin_packing.c
/**
 * Bin Packing - Heurísticas de Aproximação
 * Next Fit, First Fit, First Fit Decreasing
 * 
 * Complexidade FFD: O(n log n)
 * Fator de Aproximação FFD: 11/9 ≈ 1.222
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

#define MAX_ITEMS 100
#define CAPACITY 1.0

typedef struct {
    double capacity;
    double used;
    int items[MAX_ITEMS];
    int n_items;
} Bin;

// Função de comparação para qsort (ordem decrescente)
int compare_desc(const void* a, const void* b) {
    double diff = *(double*)b - *(double*)a;
    return (diff > 0) - (diff < 0);
}

// Next Fit: Tenta bin atual, se não couber abre novo
int next_fit(double items[], int n, Bin bins[]) {
    int n_bins = 1;
    bins[0].used = 0;
    bins[0].n_items = 0;
    
    for (int i = 0; i < n; i++) {
        if (bins[n_bins-1].used + items[i] <= CAPACITY) {
            bins[n_bins-1].items[bins[n_bins-1].n_items++] = i;
            bins[n_bins-1].used += items[i];
        } else {
            bins[n_bins].used = items[i];
            bins[n_bins].n_items = 1;
            bins[n_bins].items[0] = i;
            n_bins++;
        }
    }
    return n_bins;
}

// First Fit: Coloca no primeiro bin que cabe
int first_fit(double items[], int n, Bin bins[]) {
    int n_bins = 0;
    
    for (int i = 0; i < n; i++) {
        bool placed = false;
        
        for (int j = 0; j < n_bins; j++) {
            if (bins[j].used + items[i] <= CAPACITY) {
                bins[j].items[bins[j].n_items++] = i;
                bins[j].used += items[i];
                placed = true;
                break;
            }
        }
        
        if (!placed) {
            bins[n_bins].used = items[i];
            bins[n_bins].n_items = 1;
            bins[n_bins].items[0] = i;
            n_bins++;
        }
    }
    return n_bins;
}

// First Fit Decreasing: Ordena decrescente + First Fit
int first_fit_decreasing(double items[], int n, Bin bins[]) {
    double sorted[MAX_ITEMS];
    for (int i = 0; i < n; i++) sorted[i] = items[i];
    
    qsort(sorted, n, sizeof(double), compare_desc);
    return first_fit(sorted, n, bins);
}

void print_bins(Bin bins[], int n_bins) {
    for (int i = 0; i < n_bins; i++) {
        printf("Bin %d (%.2f/%.2f)\n", i+1, bins[i].used, CAPACITY);
    }
}

int main() {
    printf("╔═══════════════════════════════════════╗\n");
    printf("║  BIN PACKING - Heurísticas            ║\n");
    printf("╚═══════════════════════════════════════╝\n\n");
    
    double items[] = {0.7, 0.5, 0.3, 0.4, 0.2, 0.6, 0.1, 0.8};
    int n = 8;
    
    printf("Itens: ");
    for (int i = 0; i < n; i++) printf("%.1f ", items[i]);
    printf("\nCapacidade: %.1f\n\n", CAPACITY);
    
    Bin bins[MAX_ITEMS];
    
    // Next Fit
    for (int i = 0; i < MAX_ITEMS; i++) bins[i].n_items = 0;
    int nf = next_fit(items, n, bins);
    printf("=== Next Fit: %d bins ===\n", nf);
    print_bins(bins, nf);
    
    // First Fit
    printf("\n=== First Fit: ");
    for (int i = 0; i < MAX_ITEMS; i++) bins[i].n_items = 0;
    int ff = first_fit(items, n, bins);
    printf("%d bins ===\n", ff);
    print_bins(bins, ff);
    
    // First Fit Decreasing
    printf("\n=== First Fit Decreasing: ");
    for (int i = 0; i < MAX_ITEMS; i++) bins[i].n_items = 0;
    int ffd = first_fit_decreasing(items, n, bins);
    printf("%d bins ===\n", ffd);
    print_bins(bins, ffd);
    
    printf("\nResumo:\n");
    printf("NF: %d bins (fator 2)\n", nf);
    printf("FF: %d bins (fator 1.7)\n", ff);
    printf("FFD: %d bins (fator 1.222)\n", ffd);
    
    return 0;
}

File: src/21-algoritmos-aproximacao/04-bin-packing/bin_packing.py
"""
Bin Packing - Heurísticas de Aproximação
Next Fit, First Fit, First Fit Decreasing
"""

def next_fit(items, capacity=1.0):
    """Next Fit: O(n), fator 2"""
    bins = [[]]
    for item in items:
        if sum(bins[-1]) + item <= capacity:
            bins[-1].append(item)
        else:
            bins.append([item])
    return bins

def first_fit(items, capacity=1.0):
    """First Fit: O(n²), fator 1.7"""
    bins = []
    for item in items:
        placed = False
        for bin in bins:
            if sum(bin) + item <= capacity:
                bin.append(item)
                placed = True
                break
        if not placed:
            bins.append([item])
    return bins

def first_fit_decreasing(items, capacity=1.0):
    """FFD: O(n log n), fator 1.222"""
    sorted_items = sorted(items, reverse=True)
    return first_fit(sorted_items, capacity)

def main():
    print("╔═══════════════════════════════════════╗")
    print("║  BIN PACKING - Heurísticas (Python)   ║")
    print("╚═══════════════════════════════════════╝\n")
    
    items = [0.7, 0.5, 0.3, 0.4, 0.2, 0.6, 0.1, 0.8]
    capacity = 1.0
    
    print(f"Itens: {items}")
    print(f"Capacidade: {capacity}\n")
    
    nf = next_fit(items, capacity)
    print(f"=== Next Fit: {len(nf)} bins ===")
    for i, bin in enumerate(nf, 1):
        print(f"Bin {i}: {bin} = {sum(bin):.2f}")
    
    ff = first_fit(items, capacity)
    print(f"\n=== First Fit: {len(ff)} bins ===")
    for i, bin in enumerate(ff, 1):
        print(f"Bin {i}: {bin} = {sum(bin):.2f}")
    
    ffd = first_fit_decreasing(items, capacity)
    print(f"\n=== First Fit Decreasing: {len(ffd)} bins ===")
    for i, bin in enumerate(ffd, 1):
        print(f"Bin {i}: {bin} = {sum(bin):.2f}")
    
    print(f"\nResumo:")
    print(f"NF: {len(nf)} bins (fator 2)")
    print(f"FF: {len(ff)} bins (fator 1.7)")
    print(f"FFD: {len(ffd)} bins (fator 1.222)")

if __name__ == "__main__":
    main()
